File: ui/src/lib/policy.ts
/**
 * What the dashboard may offer, answered from the API's own enforcement table.
 *
 * The table itself is generated — `./policy.generated`, written by
 * `hack/gen-ui-policy` from `internal/api/policy.go` — so that the rules the
 * dashboard renders against and the rules the API enforces cannot be two
 * different opinions. This file is the other half of that split, and it is
 * hand-written on purpose: generated data plus a small helper that reasons
 * about it, rather than generated logic nobody can read.
 *
 * The question a screen asks is never "may I call `PATCH
 * /api/v1/environments/{name}`" — it is "may this person redeploy". So the
 * route is named once, at the control, and everything else comes from the
 * table: whether to render it, and the words to explain it when it is
 * disabled. Those words are the table's own `doing`, which is what the API's
 * 403 is built from, so a disabled button and the refusal behind it say the
 * same thing.
 *
 * **`may` answers admission, not completeness.** Two of the five requirement
 * kinds admit any valid token and narrow what comes back — the cross-project
 * reads to the caller's own projects, and `GET /status` to the caller's
 * platform role. There is nothing to hide for those, and no button to disable;
 * what the answer leaves out is the handler's business and cannot be read off
 * the table. `narrowsAnswer` marks them, and a screen showing one still has to
 * treat an absent field as "you may not know" rather than as a zero.
 */

import { PLATFORM_ROLES, POLICY, PROJECT_ROLES } from "./policy.generated";
import type { PlatformRole, ProjectRole, Requirement, Route } from "./policy.generated";

export { PLATFORM_ROLES, POLICY, PROJECT_ROLES, REQUIREMENT_KINDS } from "./policy.generated";
export type { PlatformRole, ProjectRole, Requirement, RequirementKind, Route } from "./policy.generated";

/**
 * Who is asking, in the two words the API tells the dashboard about them:
 * `Me.platformRole`, and the `role` that travels on every Project the caller
 * can see.
 *
 * Both are optional and both are plain strings, because both arrive from the
 * network. Anything that is not a role the table knows about — an older
 * operator's spelling, a payload that has not loaded yet — satisfies nothing
 * at all, which is `AtLeast`'s rule in `internal/access` and the safe
 * direction: a control that fails to render is a nuisance, one that renders
 * for somebody the API will refuse is a lie.
 */
export interface Caller {
  /** `Me.platformRole`: "operator" or "member". */
  platform?: string;
  /** `Project.role` on the project the control acts on: "admin", "developer" or "viewer". */
  project?: string;
  /** The project's name, used only to word a refusal the way the API words it. */
  projectName?: string;
}

/** Where a role sits in its ordering, and -1 for one this build has never
 * heard of — which is below every real role rather than equal to the weakest. */
function rank(roles: readonly string[], role: string | undefined): number {
  return role === undefined ? -1 : roles.indexOf(role);
}

/** Whether `held` carries at least the authority of `want`, comparing in the
 * order the API compares in (the arrays are generated from the same iota
 * blocks `internal/access` compares with). */
export function platformAtLeast(held: string | undefined, want: PlatformRole): boolean {
  const holds = rank(PLATFORM_ROLES, held);
  return holds >= 0 && holds >= rank(PLATFORM_ROLES, want);
}

/** The same, for a role on one project. */
export function projectAtLeast(held: string | undefined, want: ProjectRole): boolean {
  const holds = rank(PROJECT_ROLES, held);
  return holds >= 0 && holds >= rank(PROJECT_ROLES, want);
}

/**
 * The project role this caller effectively holds.
 *
 * An operator holds admin on every project, present and future — the rule
 * lives in `access.ProjectRoleFor` and is applied here for the same reason it
 * is applied there rather than in each caller. The API already says so on the
 * payload (`Project.role` reads "admin" for an operator, listed or not), so
 * this only matters where the dashboard is deciding about a project it has not
 * loaded.
 */
export function effectiveProjectRole(caller: Caller): string | undefined {
  if (platformAtLeast(caller.platform, "operator")) return "admin";
  return caller.project;
}

/** What a route asks of its caller. */
export function requirementFor(route: Route): Requirement {
  return POLICY[route];
}

/**
 * Whether a route admits this caller — the question a control asks before it
 * renders.
 *
 * It is the API's own decision, minus the part that needs the cluster: the
 * table says *which* role a route wants and the caller says which they hold,
 * but which project a request turns out to be about is resolved server-side
 * from the object it names. So the answer is only as right as the role handed
 * in; a screen acting on a project must pass that project's own `role`.
 */
export function may(route: Route, caller: Caller = {}): boolean {
  const requirement = POLICY[route];
  switch (requirement.kind) {
    case "authenticated":
    case "visibleProjects":
    case "roleShapedBody":
      // A valid token is the whole requirement. Whether the answer is narrowed
      // is a different question — see narrowsAnswer.
      return true;
    case "operator":
      return platformAtLeast(caller.platform, "operator");
    case "projectRole":
      return projectAtLeast(effectiveProjectRole(caller), requirement.role ?? "admin");
    default:
      return unknownKind(requirement.kind);
  }
}

/**
 * Why not, in the words the API refuses in — or undefined when there is
 * nothing to explain because the caller may.
 *
 * It is the same vocabulary rather than the same sentence: the API's 403 is
 * the authority on what happened, and this is what a disabled control says
 * before anybody calls it.
 */
export function refusal(route: Route, caller: Caller = {}): string | undefined {
  if (may(route, caller)) return undefined;
  const requirement = POLICY[route];
  const doing = requirement.doing ?? "this";

  if (requirement.kind === "operator") {
    const held = rank(PLATFORM_ROLES, caller.platform) >= 0 ? `; you are a ${caller.platform}` : "";
    return `${doing} needs the operator role${held}`;
  }

  const role = requirement.role ?? "admin";
  const held = effectiveProjectRole(caller);
  const on = caller.projectName ? ` on ${caller.projectName}` : "";
  if (rank(PROJECT_ROLES, held) < 0) {
    // No role at all. The API answers such a request with a not-found rather
    // than a refusal — a caller who may not see a project is not told it
    // exists — so the dashboard should not have offered the object either.
    return `${doing} needs ${role}; you have no role${on}`;
  }
  return `you have ${held}${on}; ${doing} needs ${role}`;
}

/**
 * Whether a route admits everybody and narrows what it answers with, rather
 * than admitting or refusing the call.
 *
 * A screen showing one of these is showing the caller's own slice of
 * something, and absent fields mean "you may not know" rather than zero. What
 * exactly is left out is the handler's, and is the one part of the policy this
 * table cannot state.
 */
export function narrowsAnswer(route: Route): boolean {
  const kind = POLICY[route].kind;
  return kind === "visibleProjects" || kind === "roleShapedBody";
}

/**
 * A requirement kind nothing above handles.
 *
 * The `never` parameter is the point: a kind added to the API's table and
 * regenerated into `RequirementKind` stops this file compiling, rather than
 * becoming a route the dashboard quietly treats as one anybody may call.
 */
function unknownKind(kind: never): never {
  throw new Error(`unknown requirement kind: ${String(kind)}`);
}

File: internal/signals/policy.go
/*
Copyright 2026.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package signals

import (
	"fmt"
	"sort"
	"strings"
	"time"

	kitchenv1alpha1 "github.com/Bermos/Kitchen/api/v1alpha1"
)

// The clock, as opposed to the catalogue.
//
// docs/OBSERVABILITY.md §7 said the thresholds were constants and §9 recorded
// the trade-off — "signals as versioned code vs. user-configurable rules —
// code, until the alerting era forces the question with real requirements".
// The homelab installation is that requirement: a threshold of three
// correlated projects is a threshold a three-project estate never reaches, and
// a detector that can never fire is worse than one that fires too often,
// because it reads as health.
//
// So a bounded, named set of numbers becomes configuration, and nothing else
// does. Which signals exist, what they compute and their base tier stay
// compiled in, for the reason [Signal.Version] exists: two installations on
// catalogue v1 that disagreed about what a rule *is* would make a version
// number meaningless. What they may disagree about is how impatient they are,
// and every finding records the numbers it was evaluated against so that
// `v1 @ correlatedProjects=2` and `v1 @ correlatedProjects=3` are
// distinguishable years later. That is [Policy.Provenance], and an audit pack
// that could not say what the floor was at the time would not be evidence.

// Policy is the bounded set of thresholds an installation may set.
//
// It is a plain value with no pointers and no absent fields: every caller gets
// a whole answer, and the resolution from a possibly-empty spec happens once,
// in [PolicyFrom]. A zero Policy is not "the defaults" — it is a Policy nobody
// resolved, which [Policy.Normalised] repairs and which the rules never see,
// because [Gather] always resolves one.
type Policy struct {
	// Preset is the named base the numbers came from, kept so that a screen
	// can say "balanced, with the correlation threshold moved" rather than
	// only showing six numbers.
	Preset PresetName

	// CorrelatedProjects is how many projects must be degrading together
	// before it is one platform problem rather than several application
	// problems. It replaces the [CorrelatedProjects] constant.
	CorrelatedProjects int

	// CorrelationWindow is how far apart two conditions may have started and
	// still count as the same moment.
	//
	// This is a knob the code did not have: the two cross-project detectors
	// reused [RecentWindow], which is what "now" means for a series and is
	// not the same question. The `balanced` preset is that number exactly, so
	// nothing moves for an installation that sets nothing.
	CorrelationWindow time.Duration

	// EscalationWindow is how long an owner-tier condition may sit
	// unacknowledged before the operator is added to it, and UntendedMultiple
	// how many of those windows it survives before it becomes a line on the
	// compliance posture. They replace the constants of the same names.
	EscalationWindow time.Duration
	UntendedMultiple int

	// MaxSilence is the longest silence a member may set on their own row.
	MaxSilence time.Duration

	// Paging is whether [TierPage] is delivered as a page at all. False holds
	// every paging condition down to a ticket — the homelab reading, where
	// nobody is on call.
	//
	// It is installation-wide, like every number here. #472 argues a project
	// should be able to tighten its own rows back up; #519 is that, and until
	// it lands this setting is the whole answer for every project.
	Paging bool
}

// PresetName is one of the three named bases.
type PresetName string

const (
	// PresetStrict wants to hear about it early.
	PresetStrict PresetName = "strict"
	// PresetBalanced is the platform's own judgement, and is the compiled-in
	// constants exactly.
	PresetBalanced PresetName = "balanced"
	// PresetHomelab is one host and a handful of projects.
	PresetHomelab PresetName = "homelab"
)

// presets are the three, and `balanced` is load-bearing: it is the numbers
// this package compiled in before any of this was configurable, so an
// installation that has never opened the screen behaves exactly as it did.
// A change to a value here changes what an existing installation sees.
var presets = map[PresetName]Policy{
	PresetStrict: {
		Preset:             PresetStrict,
		CorrelatedProjects: 2,
		CorrelationWindow:  30 * time.Minute,
		EscalationWindow:   30 * time.Minute,
		UntendedMultiple:   2,
		MaxSilence:         7 * 24 * time.Hour,
		Paging:             true,
	},
	PresetBalanced: {
		Preset:             PresetBalanced,
		CorrelatedProjects: CorrelatedProjects,
		CorrelationWindow:  RecentWindow,
		EscalationWindow:   EscalationWindow,
		UntendedMultiple:   UntendedMultiple,
		MaxSilence:         MaxSilence,
		Paging:             true,
	},
	PresetHomelab: {
		Preset:             PresetHomelab,
		CorrelatedProjects: 2,
		CorrelationWindow:  time.Hour,
		EscalationWindow:   4 * time.Hour,
		UntendedMultiple:   6,
		MaxSilence:         30 * 24 * time.Hour,
		Paging:             false,
	},
}

// PresetDescriptions is what each preset is for, in one sentence, served to
// the screen rather than written on it — a dashboard that spelled these out
// itself would be a second copy of a decision this package owns.
//
// The homelab sentence says what the preset does and stops there. #472 argues
// that a project should be able to tighten what *it* hears — turning paging
// back on for its own rows costs the operator nothing — and that is true and
// not built: the policy is installation-wide, and #519 is the project-scoped
// override. A description promising a control the platform does not have would
// be the worst kind of copy, since the person reading it is the one who would
// go looking for the switch.
var PresetDescriptions = map[PresetName]string{
	PresetStrict: "Two projects are a correlation, half an hour unmitigated escalates to the operator, " +
		"and a silence lasts a week.",
	PresetBalanced: "The platform's own judgement, and what every installation has had until now: three " +
		"projects correlate, an hour unmitigated escalates, a silence lasts a month.",
	PresetHomelab: "One host and a handful of projects: two projects are a correlation, nothing escalates " +
		"before the afternoon, and nothing pages — everything that would say “act now” is a ticket.",
}

// Presets lists the three in the order the screen offers them: loudest first,
// which is also the order they read in.
func Presets() []Policy {
	ordered := make([]Policy, 0, len(presets))
	for _, name := range []PresetName{PresetStrict, PresetBalanced, PresetHomelab} {
		ordered = append(ordered, presets[name])
	}
	return ordered
}

// Preset is one named base.
func Preset(name PresetName) (Policy, bool) {
	policy, ok := presets[name]
	return policy, ok
}

// DefaultPolicy is what an installation that has configured nothing runs
// under, and it is `balanced` — which is to say, the constants.
func DefaultPolicy() Policy { return presets[PresetBalanced] }

// PolicyFrom resolves the singleton's spec into the whole value the rules
// read: the named preset, with each field somebody set overriding that one
// number.
//
// A nil Kitchen, an empty spec and an unknown preset all resolve to
// [DefaultPolicy]. That is deliberate rather than defensive: the gatherer
// resolves a policy even when the singleton could not be read, and a round
// evaluated against no policy at all would be a round with no thresholds.
func PolicyFrom(kitchen *kitchenv1alpha1.Kitchen) Policy {
	if kitchen == nil {
		return DefaultPolicy()
	}
	return PolicyFromSpec(kitchen.Spec.Observability.Signals.Policy)
}

// PolicyFromSpec is [PolicyFrom]'s half that a caller holding only the spec
// can use — the API's PATCH handler, which resolves what a request would mean
// before it writes it.
func PolicyFromSpec(spec kitchenv1alpha1.SignalPolicySpec) Policy {
	policy, ok := presets[PresetName(spec.Preset)]
	if !ok {
		policy = DefaultPolicy()
	}
	if spec.CorrelatedProjects != nil {
		policy.CorrelatedProjects = int(*spec.CorrelatedProjects)
	}
	if spec.CorrelationWindowMinutes != nil {
		policy.CorrelationWindow = time.Duration(*spec.CorrelationWindowMinutes) * time.Minute
	}
	if spec.EscalationWindowMinutes != nil {
		policy.EscalationWindow = time.Duration(*spec.EscalationWindowMinutes) * time.Minute
	}
	if spec.UntendedMultiple != nil {
		policy.UntendedMultiple = int(*spec.UntendedMultiple)
	}
	if spec.MaxSilenceHours != nil {
		policy.MaxSilence = time.Duration(*spec.MaxSilenceHours) * time.Hour
	}
	if spec.Paging != nil {
		policy.Paging = *spec.Paging
	}
	return policy.Normalised()
}

// Normalised replaces anything unset or nonsensical with the default's answer
// for that one field.
//
// It exists because a Policy reaches the rules from three directions — the
// spec, a test's struct literal, and a caller that built one by hand — and a
// correlation threshold of zero would make every pair of failures an incident.
// Repairing one field at a time rather than rejecting the whole value is the
// right failure mode here: this is an observability capability, and refusing to
// evaluate the catalogue because somebody wrote a nonsense number would be the
// platform going quiet over a setting.
//
// [Policy.Preset] is what says whether a value was resolved at all, and it is
// the only way [Policy.Paging] can be repaired: a `false` bool cannot be told
// from an unset one, so an unresolved policy — no preset named — takes the
// default's paging along with every other zero field, and a resolved one keeps
// the false somebody chose. Every resolver here names a preset; nothing else
// does.
func (p Policy) Normalised() Policy {
	base := presets[PresetBalanced]
	if _, known := presets[p.Preset]; !known {
		p.Preset = PresetBalanced
		p.Paging = base.Paging
	}
	if p.CorrelatedProjects < 2 {
		p.CorrelatedProjects = base.CorrelatedProjects
	}
	if p.CorrelationWindow <= 0 {
		p.CorrelationWindow = base.CorrelationWindow
	}
	if p.EscalationWindow <= 0 {
		p.EscalationWindow = base.EscalationWindow
	}
	if p.UntendedMultiple < 1 {
		p.UntendedMultiple = base.UntendedMultiple
	}
	if p.MaxSilence <= 0 {
		p.MaxSilence = base.MaxSilence
	}
	return p
}

// UntendedAfter is how long a condition runs unacknowledged before it becomes
// a line on the compliance posture. It is the whole of the second clock, so
// that nothing multiplies the two numbers together in more than one place.
func (p Policy) UntendedAfter() time.Duration {
	return time.Duration(p.UntendedMultiple) * p.EscalationWindow
}

// Deliver is what this installation actually does with a rule's declared tier.
//
// The only thing a policy may do to a tier is hold a page down to a ticket,
// and only when paging is off. It never raises one and never lowers past
// ticket: the base tier is catalogue knowledge (#471, decision 4), and a
// policy that could rewrite it would be the configurable-rules door this
// package spent §9 keeping shut.
func (p Policy) Deliver(tier Tier) Tier {
	if tier == TierPage && !p.Paging {
		return TierTicket
	}
	return tier
}

// Matches reports whether a policy is one of the presets exactly, which is how
// a screen tells "balanced" from "balanced, with two numbers moved".
func (p Policy) Matches(name PresetName) bool {
	preset, ok := presets[name]
	if !ok {
		return false
	}
	preset.Preset = p.Preset
	return preset == p
}

// Provenance is the finding's record of what it was evaluated against: one
// short, ordered, stable string.
//
// It is one string rather than six columns because of what it is for. Nobody
// queries on the escalation window; somebody reads a finding from eight months
// ago and asks what the floor was when it fired, and an audit pack quotes the
// answer. A single field also means a threshold added later widens the string
// rather than migrating the table.
//
// The order is fixed and the spelling is durations, not seconds, because this
// is read by people.
func (p Policy) Provenance() string {
	paging := "off"
	if p.Paging {
		paging = "on"
	}
	return strings.Join([]string{
		"preset=" + string(p.Preset),
		fmt.Sprintf("correlatedProjects=%d", p.CorrelatedProjects),
		"correlationWindow=" + p.CorrelationWindow.String(),
		"escalationWindow=" + p.EscalationWindow.String(),
		fmt.Sprintf("untendedMultiple=%d", p.UntendedMultiple),
		"maxSilence=" + p.MaxSilence.String(),
		"paging=" + paging,
	}, " ")
}

// PolicyDifference names what a change moved, for the audit record and for the
// sentence the API answers with. It is empty for a write that changed nothing,
// which is what stops a no-op PATCH appending a line to the audit log.
func PolicyDifference(was, now Policy) []string {
	changed := make([]string, 0, 7)
	for _, field := range []struct {
		name       string
		was, becam string
	}{
		{"preset", string(was.Preset), string(now.Preset)},
		{"correlatedProjects", fmt.Sprint(was.CorrelatedProjects), fmt.Sprint(now.CorrelatedProjects)},
		{"correlationWindow", was.CorrelationWindow.String(), now.CorrelationWindow.String()},
		{"escalationWindow", was.EscalationWindow.String(), now.EscalationWindow.String()},
		{"untendedMultiple", fmt.Sprint(was.UntendedMultiple), fmt.Sprint(now.UntendedMultiple)},
		{"maxSilence", was.MaxSilence.String(), now.MaxSilence.String()},
		{"paging", onOff(was.Paging), onOff(now.Paging)},
	} {
		if field.was != field.becam {
			changed = append(changed, fmt.Sprintf("%s %s→%s", field.name, field.was, field.becam))
		}
	}
	sort.SliceStable(changed, func(i, j int) bool { return changed[i] < changed[j] })
	return changed
}

func onOff(on bool) string {
	if on {
		return "on"
	}
	return "off"
}
